src/sm.c

```diff
@@ -0,0 +1,480 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdlib.h>
+#include <stdint.h>
+#include <inttypes.h>
+
+#include "windows.h"
+
+#include "azure_macro_utils/macro_utils.h"
+
+#include "azure_c_util/xlogging.h"
+#include "azure_c_util/gballoc.h"
+#include "azure_c_util/interlocked_hl.h"
+
+#include "azure_c_util/sm.h"
+
+#define SM_STATE_VALUES             \
+    SM_CREATED,                     \
+    SM_OPENING,                     \
+    SM_OPENED,                      \
+    SM_OPENED_DRAINING_TO_BARRIER,  \
+    SM_OPENED_DRAINING_TO_CLOSE,    \
+    SM_OPENED_BARRIER,              \
+    SM_CLOSING                      \
+
+MU_DEFINE_ENUM(SM_STATE, SM_STATE_VALUES)
+
+#define SM_STATE_MASK       127
+#define SM_STATE_INCREMENT  256 /*at every state change, the state is incremented by this much*/
+
+#define SM_CLOSE_BIT 128
+
+#define PRI_SM_STATE "" PRI_MU_ENUM " SM_CLOSE_BIT=%d"
+
+#define SM_STATE_VALUE(state) MU_ENUM_VALUE(SM_STATE, (SM_STATE)((state) & SM_STATE_MASK)), (((state)&SM_CLOSE_BIT)>0)
+
+typedef struct SM_HANDLE_DATA_TAG
+{
+    volatile LONG state;
+    volatile LONG n; /*number of API calls to non-barriers*/
```

Comment:
>n [](start = 18, length = 1)

/poetry
Can we pretty please call this something more explicit (non_barrier_call_count)?
Please and thank you!

Reply:
Thy will be done!

---
In reply to: [423411737](https://github.com/Azure/azure-c-util/pull/37#discussion_r423411737) [](ancestors = 423411737)